### DISCUSSION TITLE
Convert Perl Hash to C++ stl Map via SWIG.

### DISCUSSION THREAD

**Discussion by shashisonam** [swig/swig#2473] [Q&A] [answered]
I am new to SWIG and I wanted to write SWIG interface where we can pass the Perl Hash and then it converts it into C++ stl map or unordered_map.
Do you have any example for that like swig interface file and then c++ file and then perl file.
I tried with those but it seems not working for me.
tried with below sample code but it seems not working for me.

example.i - swig interface file

%module example

%{
#include "example.h"
%}

%include "std_map.i"

%template(StringIntMap) std::map<std::string, int>;

%include "example.h"

example.h - C++ header file

#include <map>
#include <string>

void processMap(std::map<std::string, int> &m) {

}

perl file - test.pl
#!/opt/perl_5.18.2/bin/perl
use strict;
use warnings;
use Data::Dumper;
use lib '.';
use example;

my %hash = (
"key1" => 1,
"key2" => 2,
"key3" => 3
);

example::processMap(%hash);

the compilation and then linked with the shared object of that but it is not working , then could you please provide me the proper solution for that like we need to pass Perl Hash and then this hash will get stored into c++ stl map or unordered_map via SWIG .

**Comment by wsfulton** [comment, marked as answer]
SWIG is a tool to help marshal C++ types to Perl. You'll need to figure out the Perl API required to do what you want. If you can write that by hand, we can help with porting it into SWIG typemaps.

**Reply by shashisonam** [reply]
Hi William,
Wish you a very Happy New Year to you and your loved Ones.
As I am new to SWIG and I need help to while writing the SWIG interface.
In Perl we have a hash example below like 
my %hash = (
"key1" => "test",
"key2" => "sam",
"key3" => "perth"
);

this hash we wanted to convert into C++ stl Unordered_map or map via SWIG interface using SWIG typemaps.
Also need to pass the Perl Object to C++ Object via SWIG interface .How to do that ,
Could you please provide me some example for this .
As this is bit urgent for me.

Thanks & Regards
Shashi Singh

**Reply by shashisonam** [reply]
Hi William

**Comment by ojwb** [comment]
William has essentially answered the question, but to clarify he's talking about the Perl C API, as documented in `perldoc perlapi` (`perldoc perlxs` is also relevant).
